store/tikv/coprocessor.go

```diff
@@ -276,62 +283,55 @@ func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, desc bo
 		}
 	}
 
+	var tasks *copTask
 	if desc {
-		reverseTasks(tasks)
+		tasks = reverseTasks(&head)
+	} else {
+		tasks = head.next
 	}
 	if elapsed := time.Since(start); elapsed > time.Millisecond*500 {
-		log.Warnf("buildCopTasks takes too much time (%v), range len %v, task len %v", elapsed, rangesLen, len(tasks))
+		log.Warnf("buildCopTasks takes too much time (%v), range len %v, task len %v", elapsed, rangesLen, length)
 	}
-	txnRegionsNumHistogram.WithLabelValues("coprocessor").Observe(float64(len(tasks)))
+	txnRegionsNumHistogram.WithLabelValues("coprocessor").Observe(float64(length))
 	return tasks, nil
 }
 
-func reverseTasks(tasks []*copTask) {
-	for i := 0; i < len(tasks)/2; i++ {
-		j := len(tasks) - i - 1
-		tasks[i], tasks[j] = tasks[j], tasks[i]
-		tasks[i].idx, tasks[j].idx = tasks[j].idx, tasks[i].idx
-	}
+func reverseTasks(head *copTask) *copTask {
+	// reverse a list, head is a dummy head node.
+	p := head.next
+	head.next = nil
+	// iterator the real list and insert each node after head.
+	for p != nil {
+		save := p.next
+		p.next = head.next
+		head.next = p
+		p = save
+	}
+	return head.next
 }
 
 type copIterator struct {
 	store       *tikvStore
 	req         *kv.Request
 	concurrency int
-	mu          struct {
-		sync.RWMutex
-		tasks    []*copTask
-		respGot  int
-		finished bool
-	}
+	finished    chan struct{}
+	taskCh      chan *copTask
+
+	// If keepOrder, result is stored in copTask.respChan, tasks is a list, read them out one by one.
+	tasks *copTask
+
+	// Otherwise, result is stored in respChan.
 	respChan chan *coprocessor.Response
 	errChan  chan error
+	pending  int64
 }
 
 const minLogCopTaskTime = 300 * time.Millisecond
 
-// Pick the next new copTask and send request to tikv-server.
-func (it *copIterator) work(ctx goctx.Context) {
-	for {
-		it.mu.Lock()
-		if it.mu.finished {
-			it.mu.Unlock()
-			break
-		}
-		// Find the next task to send.
-		var task *copTask
-		for _, t := range it.mu.tasks {
-			if t.status == taskNew {
-				task = t
-				break
-			}
-		}
-		if task == nil {
-			it.mu.Unlock()
-			break
-		}
-		task.status = taskRunning
-		it.mu.Unlock()
+// The worker function that get a copTask from channel, handle it and
+// send the result back.
+func (it *copIterator) work(ctx goctx.Context, taskCh <-chan *copTask) {
+	for task := range taskCh {
 		bo := NewBackoffer(copNextMaxBackoff, ctx)
 		startTime := time.Now()
 		resp, err := it.handleTask(bo, task)
```

Comment:
The multiple responses can be sent to the original task's channel.

Instead of just read one response, the reader reads multiple responses from a task's channel until the channel is closed.

Then we don't need a task list because the original tasks never change.